File: solutions/haskell/secret-handshake/src/SecretHandshake.hs
{-# LANGUAGE BinaryLiterals #-}

module SecretHandshake (handshake) where

import Data.Bits ((.&.))
import Control.Monad.Writer

handshake :: Int -> [String]
handshake n = reverseIfNeeded n $ execWriter $ do
    when (n .&. 0b0001 > 0) $ tell ["wink"]
    when (n .&. 0b0010 > 0) $ tell ["double blink"]
    when (n .&. 0b0100 > 0) $ tell ["close your eyes"]
    when (n .&. 0b1000 > 0) $ tell ["jump"]
    where
        reverseIfNeeded n =
            if n .&. 0b10000 > 0
            then reverse
            else id

File: solutions/cpp/secret-handshake/secret_handshake.cpp
#include "secret_handshake.h"

#include <algorithm>

namespace secret_handshake {
    std::vector<std::string> commands(int n) {
        std::vector<std::string> xs;
        if (n & 0b00001) xs.push_back("wink");
        if (n & 0b00010) xs.push_back("double blink");
        if (n & 0b00100) xs.push_back("close your eyes");
        if (n & 0b01000) xs.push_back("jump");
        if (n & 0b10000) std::reverse(xs.begin(), xs.end());
        return xs;
    }
}  // namespace secret_handshake
